### ISSUE TITLE
type error

### ISSUE THREAD

**Issue by Ezreal09** [colinhacks/zod#2011]
in type.d.ts file, Some type errors occur here： Writeable<FilterEnum<T, ToExclude[number]>>
version: '3.20.3'
![image](https://user-images.githubusercontent.com/44263381/217505000-f6dd7310-410f-40e9-b1e7-5e3b02490d09.png)

**Comment by SimplyLinn** [comment]
Having done some investigation.

Issue is present in typescript 4.9.5

Removing "Writeable<>" wrapping the FilterEnum solves it for me, as the returned type from FilterEnum is already writeable.

**Comment by SimplyLinn** [comment]
I think [this PR](https://github.com/colinhacks/zod/pull/1652) introduced an issue once it's compiled to a d.ts file.

The exclude method becomes this type declaration:
```ts
declare class ZodEnum<T extends [string, ...string[]]> extends ZodType<T[number], ZodEnumDef<T>> {
    exclude<ToExclude extends readonly [T[number], ...T[number][]]>(values: ToExclude): ZodEnum<Writeable<FilterEnum<T, ToExclude[number]>>>;
}
```

Wrapping the output of FilterEnum in Writeable makes the typechecker not see it as matching the constraint set by ZodEnum.

The wrapping of Writeable, I think, is due to the order of the createZodEnum functions overloads?

I will investigate whether my assumption is correct, and potentially open a PR

**Comment by paulsmithkc** [comment]
Seeing the same error with:

zod@3.20.3
typescript@4.9.5

```
node_modules/zod/lib/types.d.ts(668,97):***error***TS2344:***Type***'Writeable<FilterEnum<T,***ToExclude[number]>>'***does***not***satisfy***the***constraint***'[string,***...string[]]'.
```

This error does not appear with the previous version of zod (3.20.2)

This same error appears for typescript versions 4.8.x and greater.

**Comment by colinhacks** [comment]
Fixed in Zod 3.20.5
